pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -0,0 +1,418 @@
+	pods, err := r.getGadgetPods(context.Background(), []string{})
+	if err != nil {
+		return nil, fmt.Errorf("get gadget pods: %w", err)
+	}
+	if len(pods) == 0 {
+		return nil, fmt.Errorf("no valid pods found to get catalog from")
+	}
+
+	pod := pods[0]
+	dialOpt := grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
+		return NewK8SExecConn(ctx, pod, time.Second*30)
+		// return NewK8SPortForwardConn(ctx, s, time.Second*30)
+	})
+
+	conn, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		panic(err)
+	}
+	client := pb.NewGadgetManagerClient(conn)
+	defer conn.Close()
+
+	info, err := client.GetInfo(ctx, &pb.InfoRequest{Version: "1.0"})
+	if err != nil {
+		return nil, fmt.Errorf("get info from gadget pod: %w", err)
+	}
+
+	catalog := &runtime.Catalog{}
+	err = json.Unmarshal(info.Catalog, &catalog)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling catalog: %w", err)
+	}
+
+	return catalog, nil
+}
+
+func (r *Runtime) storeCatalog(catalog *runtime.Catalog) error {
+	configFile, err := r.getCatalogFilename()
+	if err != nil {
+		return fmt.Errorf("get catalog filename: %w", err)
+	}
+	err = os.MkdirAll(filepath.Dir(configFile), 0o750)
+	if err != nil && errors.Is(err, os.ErrExist) {
+		return fmt.Errorf("create config dir: %w", err)
+	}
+	catalogJSON, err := json.Marshal(catalog)
+	if err != nil {
+		return fmt.Errorf("marshaling catalog JSON: %w", err)
+	}
+	err = os.WriteFile(configFile, catalogJSON, 0o644)
+	if err != nil {
+		return fmt.Errorf("write catalog file: %w", err)
+	}
+	return nil
+}
+
+func (r *Runtime) Init(runtimeGlobalParams *params.Params) error {
+	catalog, err := r.loadLocalGadgetCatalog()
+	if err == nil {
+		r.catalog = catalog
+		return nil
+	}
+
+	// fallback to embedded catalog
+	catalog = &runtime.Catalog{}
+	err = json.Unmarshal(embeddedCatalog, &catalog)
+	if err == nil {
+		r.catalog = catalog
+		return nil
+	}
+
+	// Try remote
+	catalog, err = r.loadRemoteGadgetCatalog()
+	if err != nil {
+		return fmt.Errorf("could not get catalog: %w", err)
+	}
+	err = r.storeCatalog(catalog)
+	if err != nil {
+		log.Warnf("could not store catalog: %v", err)
+	}
+	r.catalog = catalog
+	return nil
+}
+
+func (r *Runtime) Close() error {
+	return nil
+}
+
+func (r *Runtime) ParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:         ParamNode,
+			Description: "Comma-separated list of nodes to run the gadget on",
+		},
+	}
+}
+
+func (r *Runtime) GlobalParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:            ParamConnectionMethod,
+			Description:    "Method that should be used to connect to the Inspektor Gadget nodes",
+			PossibleValues: []string{"kubeapi-server-per-node", "grpc-direct"},
+			DefaultValue:   "kubeapi-server-per-node",
+		},
+	}
+}
+
+func (r *Runtime) getGadgetPods(ctx context.Context, nodes []string) ([]v1.Pod, error) {
```

Comment:
nit: perhaps it's better to create a custom structure to represent a gadget pod -> https://github.com/inspektor-gadget/inspektor-gadget/commit/032bdf7470ca8281012ef85ec90ce1dd621bee0e